docs: Add troubleshooting docs for Ingress (#20428)
cilium-team/cilium: Documentation/operations/troubleshooting_servicemesh.rst

Generic
-------

 #. Validate that the ``ds/cilium`` as well as the ``deployment/cilium-operator`` pods
    are healthy and ready.

    .. code-block:: shell-session

       $ cilium status

Manual Verification of Setup
----------------------------

 #. Validate that the ``kubeProxyReplacement`` is set to either partial or strict.

    .. code-block:: shell-session

        $ kubectl exec -n kube-system ds/cilium -- cilium status
        ...
        KVStore:                 Ok   Disabled
        Kubernetes:              Ok   1.23 (v1.23.6) [linux/amd64]
        KubeProxyReplacement:    Strict   [eth0 192.168.49.2]

 #. Validate that runtime the values of ``enable-envoy-config`` and ``enable-ingress-controller``
    are true. Ingress controller flag is optional if customer only uses ``CiliumEnvoyConfig`` or
    ``CiliumClusterwideEnvoyConfig`` CRDs.

    .. code-block:: shell-session

        $ kubectl -n kube-system get cm cilium-config -o json | egrep "enable-ingress-controller|enable-envoy-config"
                "enable-envoy-config": "true",
                "enable-ingress-controller": "true",

Ingress Troubleshooting
-----------------------

> Internally, the Cilium Ingress controller will create one Load Balancer service, one CiliumEnvoyConfig

Reviewer: ```suggestion
Internally, the Cilium Ingress controller will create one Load Balancer service, one ``CiliumEnvoyConfig``
```

and one dummy Endpoint resource for each Ingress resource.

    .. code-block:: shell-session

        $ kubectl get ingress
        NAME            CLASS    HOSTS   ADDRESS        PORTS   AGE
        basic-ingress   cilium   *       10.97.60.117   80      16m

        $ kubectl get service cilium-ingress-basic-ingress
        NAME                           TYPE           CLUSTER-IP     EXTERNAL-IP    PORT(S)        AGE
        cilium-ingress-basic-ingress   LoadBalancer   10.97.60.117   10.97.60.117   80:31911/TCP   17m

        $ kubectl get cec cilium-ingress-default-basic-ingress
        NAME                                   AGE
        cilium-ingress-default-basic-ingress   18m

 #. Validate if the Load Balancer service is having either external IP or FQDN assigned
    If it's not available after long time, please check the Load Balancer related

>     documentation from your respective cloud provider.

Reviewer: ```suggestion
 #. Validate that the Load Balancer service has either an external IP or FQDN assigned.
    If it's not available after a long time, please check the Load Balancer related
    documentation from your respective cloud provider.
```

 #. Validate if there is any warning or error message while Cilium is trying to provision
    the ``CiliumEnvoyConfig`` resource. This is unlikely to happen for CEC resources

>     originating from the Cilium Ingress controller.

Reviewer: ```suggestion
 #. Check if there is any warning or error message while Cilium is trying to provision
    the ``CiliumEnvoyConfig`` resource. This is unlikely to happen for CEC resources
    originating from the Cilium Ingress controller.
```

    .. include:: ../gettingstarted/servicemesh/warning.rst

Connectivity Troubleshooting
----------------------------

This section is for troubleshooting connectivity issues mainly for Ingress resources, but

> the same steps can be applied to manually configured CiliumEnvoyConfig resources as well.

Reviewer: ```suggestion
the same steps can be applied to manually configured ``CiliumEnvoyConfig`` resources as well.
```

> It's the best to have ``debug`` and ``debug-verbose`` enabled with below values. Kindly

Reviewer: ```suggestion
It's best to have ``debug`` and ``debug-verbose`` enabled with below values. Kindly
```

note that any change of Cilium flags requires a restart of the Cilium agent and operator.

    .. code-block:: shell-session

        $ kubectl get -n kube-system cm cilium-config -o json | grep "debug"
                "debug": "true",
                "debug-verbose": "flow",

.. note::

    The Ingress traffic is always allowed to pass through Cilium, regardless of the of

>     the related CiliumNetworkPolicy for underlying pods.

Reviewer: ```suggestion
    The Ingress traffic is always allowed to pass through Cilium, regardless of
    the related CiliumNetworkPolicy for underlying pods.
```

The request is normally traversing from LoadBalancer service to pre-assigned port of your
node, then get forwarded to the Cilium Envoy proxy, and finally get proxied to the actual

> backend service.

Reviewer: ```suggestion
The request normally traverses from LoadBalancer service to pre-assigned port of your
node, then gets forwarded to the Cilium Envoy proxy, and finally gets proxied to the actual
backend service.
```

 #. The first step between cloud Load Balancer to node port is out of Cilium scope. Please
    check related documentation from your respective cloud provider to make sure your
    clusters are configured properly.

 #. The second step could be checked by ssh to your underlying host, and send the similar

>     request to localhost.

Reviewer: ```suggestion
 #. The second step could be checked by connecting with SSH to your underlying host, and sending a
    request to localhost on the relevant port:
```

    .. code-block:: shell-session

        $ kubectl get service cilium-ingress-basic-ingress
        NAME                           TYPE           CLUSTER-IP     EXTERNAL-IP    PORT(S)        AGE
        cilium-ingress-basic-ingress   LoadBalancer   10.97.60.117   10.97.60.117   80:31911/TCP   17m

        # After ssh to any of k8s node
        $ curl -v http://localhost:31911/
        *   Trying 127.0.0.1:31911...
        * TCP_NODELAY set
        * Connected to localhost (127.0.0.1) port 31911 (#0)
        > GET / HTTP/1.1
        > Host: localhost:31911
        > User-Agent: curl/7.68.0
        > Accept: */*
        >
        * Mark bundle as not supporting multiuse
        < HTTP/1.1 503 Service Unavailable
        < content-length: 19
        < content-type: text/plain
        < date: Thu, 07 Jul 2022 12:25:56 GMT
        < server: envoy
        <
        * Connection #0 to host localhost left intact

        # Flows for world identity
        $ kubectl -n kube-system exec ds/cilium -- hubble observe -f --identity 2
        Jul  7 12:28:27.970: 127.0.0.1:54704 <- 127.0.0.1:13681 http-response FORWARDED (HTTP/1.1 503 0ms (GET http://localhost:31911/))

>     Alternatively, you can also send request directly to Envoy proxy port. For Ingress, the proxy port is randomly

Reviewer: ```suggestion
    Alternatively, you can also send a request directly to the Envoy proxy port. For Ingress, the proxy port is randomly
```

    assigned by the Cilium Ingress controller. For manually configured ``CiliumEnvoyConfig`` resources, the proxy
    port is retrieved directly from the spec.

    .. code-block:: shell-session

>         $  kubectl logs -f -n kube-system ds/cilium --timestamps| egrep "envoy|proxy"

Reviewer: ```suggestion
        $  kubectl logs -f -n kube-system ds/cilium --timestamps | egrep "envoy|proxy"
```

        ...
        2022-07-08T08:05:13.986649816Z level=info msg="Adding new proxy port rules for cilium-ingress-default-basic-ingress:19672" proxy port name=cilium-ingress-default-basic-ingress subsys=proxy

        # After ssh to any of k8s node, send request to Envoy proxy port directly
        $ curl -v  http://localhost:19672
        *   Trying 127.0.0.1:19672...
        * TCP_NODELAY set
        * Connected to localhost (127.0.0.1) port 19672 (#0)
        > GET / HTTP/1.1
        > Host: localhost:19672
        > User-Agent: curl/7.68.0
        > Accept: */*
        >
        * Mark bundle as not supporting multiuse
        < HTTP/1.1 503 Service Unavailable
        < content-length: 19
        < content-type: text/plain
        < date: Fri, 08 Jul 2022 08:12:35 GMT
        < server: envoy

>     If you see the response similar to above, it means that the request is being

Reviewer: ```suggestion
    If you see a response similar to the above, it means that the request is
```

    redirected to proxy successfully. The http response will have one special header
    ``server: envoy`` accordingly. The same can be observed from ``hubble observe``
    command :ref:`hubble_troubleshooting`.

    The most common root cause is either that the Cilium Envoy proxy is not running
    on the node, or there is some other issue with CEC resource provisioning.

    .. code-block:: shell-session

        $ kubectl exec -n kube-system ds/cilium -- cilium status
        ...
        Controller Status:       49/49 healthy
        Proxy Status:            OK, ip 10.0.0.25, 6 redirects active on ports 10000-20000
        Global Identity Range:   min 256, max 65535

 #. Assume that the above steps are done successfully, you can either send requests either
    via external IP/FQDN In order to check if the request is proxied to backend service

>     correctly.

Reviewer: “Assuming”, and this sentence is not correct, please fix it

    Please double check if your backend service is up and healthy. The Envoy Discovery Service

>     (EDS) is having name following the convention ``<namespace>/<service-name>:<port>``.

Reviewer: ```suggestion
    Double-check whether your backend service is up and healthy. The Envoy Discovery Service
    (EDS) has a name that follows the convention ``<namespace>/<service-name>:<port>``.
```

    .. code-block:: shell-session

>         $ LB_IP=$(kg ingress basic-ingress -o json | jq '.status.loadBalancer.ingress[0].ip' | jq -r .)

Reviewer 2: small nit: `kg` --> `kubectl get`
